pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java

```diff
@@ -60,196 +56,151 @@
  * </ul>
  */
 public final class OffHeapBitmapInvertedIndexCreator implements InvertedIndexCreator {
-  // Maximum number of values due to 2GB file size limit
-  public static final int MAX_NUM_VALUES = Integer.MAX_VALUE / V1Constants.Numbers.INTEGER_SIZE;
-
-  private final String _columnName;
+  private final File _invertedIndexFile;
   private final boolean _singleValue;
   private final int _cardinality;
   private final int _numDocs;
   private final int _numValues;
-  private final File _indexDir;
-  private final File _invertedIndexFile;
 
-  private ByteBuffer _valueByteBuffer;
-  private IntBuffer _valueIntBuffer;
-
-  // For multi-value column only
-  private int _valueBufferOffset;
-  private ByteBuffer _numValuesByteBuffer;
-  private IntBuffer _numValuesIntBuffer;
-
-  // Posting list related buffers
-  // Constructed in first pass
-  private ByteBuffer _postingListLengthByteBuffer;
-  private IntBuffer _postingListLengthIntBuffer;
-  // Constructed in second pass
-  private ByteBuffer _postingListValueByteBuffer;
-  private ByteBuffer _postingListCurrentOffsetByteBuffer;
+  // For single-valued and multi-valued column
+  private int _nextDocId;
+  private LBuffer _valueBuffer;
+  private LBuffer _postingListIndexBuffer;
+  private LBuffer _postingListValueBuffer;
+  // For multi-valued column only
+  private int _nextValueId;
+  private LBuffer _numValuesBuffer;
 
   public OffHeapBitmapInvertedIndexCreator(File indexDir, FieldSpec fieldSpec, int cardinality, int numDocs,
       int numValues) {
-    _columnName = fieldSpec.getName();
+    _invertedIndexFile =
+        new File(indexDir, fieldSpec.getName() + V1Constants.Indexes.BITMAP_INVERTED_INDEX_FILE_EXTENSION);
     _singleValue = fieldSpec.isSingleValueField();
     _cardinality = cardinality;
     _numDocs = numDocs;
-    if (_singleValue) {
-      Preconditions.checkArgument(numDocs <= MAX_NUM_VALUES,
-          "For single-value column: %s, numDocs: %s is too large to create inverted index", _columnName, numDocs);
-      _numValues = numDocs;
-    } else {
-      Preconditions.checkArgument(numValues <= MAX_NUM_VALUES,
-          "For multi-value column: %s, numValues: %s is too large to create inverted index", _columnName, numValues);
-      _numValues = numValues;
-    }
-    _indexDir = indexDir;
-    _invertedIndexFile = new File(indexDir, _columnName + V1Constants.Indexes.BITMAP_INVERTED_INDEX_FILE_EXTENSION);
+    _numValues = _singleValue ? numDocs : numValues;
 
     try {
-      // Create value buffer
-      _valueByteBuffer = MmapUtils.allocateDirectByteBuffer(_numValues * V1Constants.Numbers.INTEGER_SIZE, null,
-          "bitmap value buffer for: " + _columnName);
-      _valueIntBuffer = _valueByteBuffer.asIntBuffer();
-
-      // Create numValues buffer for multi-value column
+      _valueBuffer = new LBuffer((long) _numValues * Integer.BYTES);
```

Comment:
Can you try using larray? It supports mmapping the files. With mmap, we don't have to limit the number of inv index created, or somehow make sure that the jvm is started with the max amount of memory for all inv indices to be created in parallel.

Alternativey (safer, tested) can you create a list of PinotDataBuffer objects? It may be a bit more complex in implementation, but will save a lot in operational cost. Speed is not a concern here, so we should be good.

Reply:
Actually we do care about performance on server side, or segment load could take forever. On non-SSD machine, using MMapBuffer could be very slow because we need to write the whole buffer to the disk. We have been using direct memory buffer for inverted index generation, so I'll keep that behavior for now.

Reply:
Add a threshold (25M values) so that the creator can choose to use LBuffer or MMapBuffer automatically.